compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp

```diff
@@ -0,0 +1,424 @@
+findDominatedByAllOps(const SetVector<Operation *> &ops,
+                      const DominanceInfo &dominanceInfo) {
+  for (auto opA : ops) {
+    bool opAIsDominatedByAll = true;
+    for (auto opB : ops) {
+      if (opA == opB) {
+        continue;
+      }
+      if (!dominanceInfo.properlyDominates(opB, opA)) {
+        opAIsDominatedByAll = false;
+        break;
+      }
+    }
+    if (opAIsDominatedByAll) {
+      return opA;
+    }
+  }
+  return std::nullopt;
+}
+
+/// Find all candidates that can be used for horizontal fusion
+static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+    linalg::LinalgOp seedOp, std::optional<linalg::GenericOp> truncOp,
+    const llvm::SmallDenseSet<linalg::LinalgOp> &groupedOperations,
+    const DominanceInfo &dominanceInfo) {
+
+  SetVector<Operation *> allOps;
+  SmallVector<linalg::LinalgOp> contractionOps = {seedOp};
+  std::optional<SmallVector<linalg::GenericOp>> truncateOps;
+  if (truncOp) {
+    truncateOps = {truncOp.value()};
+  }
+  allOps.insert(seedOp);
+  if (truncOp) {
+    allOps.insert(truncOp.value());
+  }
+
+  Value lhs = seedOp->getOperand(0);
+  Type lhsType = lhs.getType();
+  Value rhs = seedOp->getOperand(1);
+  Type rhsType = rhs.getType();
+  Value out = seedOp->getOperand(2);
+  Type outType = out.getType();
+
+  auto canBeGrouped = [&](Operation *op) -> bool {
+    if (op->getParentOp() != seedOp->getParentOp()) {
+      return false;
+    }
+    auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+    if (!linalgOp) {
+      return false;
+    }
+    if (!linalg::isaContractionOpInterface(linalgOp)) {
+      return false;
+    }
+    if (groupedOperations.contains(linalgOp) || allOps.contains(linalgOp)) {
+      return false;
+    }
+    if (linalgOp->getOperand(0).getType() != lhsType ||
+        linalgOp->getOperand(1).getType() != rhsType ||
+        linalgOp->getOperand(2).getType() != outType) {
+      return false;
+    }
+    // Either this has to dominate the seed or seed has to dominate this.
+    if (!dominanceInfo.properlyDominates(linalgOp, seedOp) &&
+        !dominanceInfo.properlyDominates(seedOp, linalgOp)) {
+      return false;
+    }
+    return true;
+  };
+
+  // Iterating over the users as is bad cause the users have no ordering
+  // gaurantees. So look for only ops within the same block as the seed op and
+  // sort them.
+  SmallVector<Operation *> lhsUsers;
+  for (Operation *lhsUser : lhs.getUsers()) {
+    if (lhsUser->getBlock() == seedOp->getBlock()) {
+      lhsUsers.push_back(lhsUser);
+    }
+  }
+  llvm::sort(lhsUsers, [&](Operation *lhs, Operation *rhs) {
+    return dominanceInfo.properlyDominates(lhs, rhs);
+  });
+
+  // Collect all contraction op users of lhs.
+  for (Operation *lhsUser : lhsUsers) {
+    if (lhsUser == seedOp.getOperation()) {
+      continue;
+    }
+
+    auto linalgUser = dyn_cast<linalg::LinalgOp>(lhsUser);
+    if (!canBeGrouped(linalgUser)) {
+      continue;
+    }
```

Comment:
IMO `canBeGrouped` should directly take `linalg::LinalgOp` inputs (not `Operation*`). Then you don't need the casting check in the lambda. And here can be:

```
if (!linalgUser || !canBeGrouped(linalgUser)) {
  continue;
}
```